autogpt/memory/vector/memory_item.py

```diff
@@ -0,0 +1,223 @@
+                question=question_for_summary,
+            )[0]
+        )
+        logger.debug("Total summary: " + summary)
+
+        # TODO: investigate search performance of weighted average vs summary
+        # e_average = np.average(e_chunks, axis=0, weights=[len(c) for c in chunks])
+        e_summary = get_embedding(summary)
+
+        metadata["source_type"] = source_type
+
+        return MemoryItem(
+            text,
+            summary,
+            chunks,
+            chunk_summaries,
+            e_summary,
+            e_chunks,
+            metadata=metadata,
+        )
+
+    @staticmethod
+    def from_text_file(content: str, path: str):
+        return MemoryItem.from_text(content, "text_file", {"location": path})
+
+    @staticmethod
+    def from_code_file(content: str, path: str):
+        # TODO: implement tailored code memories
+        return MemoryItem.from_text(content, "code_file", {"location": path})
+
+    @staticmethod
+    def from_ai_action(ai_message: Message, result_message: Message):
+        # The result_message contains either user feedback
+        # or the result of the command specified in ai_message
+
+        if ai_message["role"] != "assistant":
+            raise ValueError(f"Invalid role on 'ai_message': {ai_message['role']}")
+
+        result = (
+            result_message["content"]
+            if result_message["content"].startswith("Command")
+            else "None"
+        )
+        user_input = (
+            result_message["content"]
+            if result_message["content"].startswith("Human feedback")
+            else "None"
+        )
+        memory_content = (
+            f"Assistant Reply: {ai_message['content']}"
+            "\n\n"
+            f"Result: {result}"
+            "\n\n"
+            f"Human Feedback: {user_input}"
+        )
+
+        return MemoryItem.from_text(
+            text=memory_content,
+            source_type="agent_history",
+            how_to_summarize="if possible, also make clear the link between the command in the assistant's response and the command result. Do not mention the human feedback if there is none",
+        )
+
+    @staticmethod
+    def from_webpage(content: str, url: str, question: str | None = None):
+        return MemoryItem.from_text(
+            text=content,
+            source_type="webpage",
+            metadata={"location": url},
+            question_for_summary=question,
+        )
+
+    def __str__(self) -> str:
+        token_length = count_string_tokens(self.raw_content, Config().embedding_model)
+        return f"""
+=============== MemoryItem ===============
+Length: {token_length} tokens in {len(self.e_chunks)} chunks
+Metadata: {json.dumps(self.metadata, indent=2)}
+---------------- SUMMARY -----------------
+{self.summary}
+------------------ RAW -------------------
+{self.raw_content}
+==========================================
+"""
+
+
+@dataclasses.dataclass
+class MemoryItemRelevance:
+    """
```

Comment:
Why isn't the the domain of the vector memory provider?  Different backends are going to provide different ways of looking up relevant items (often in db).

Reply:
They will surely provide different ways of looking up relevant items, but I aim to have a homogeneous interface for the selected and scored items. It's easy to do for the `JSONFileMemory` but also works fine with my new `RedisMemory` implementation.

Reply:
Okay.  I'll want to talk with you about this in more detail, but fine for now.